### PR TITLE
chore(declarative) on the fly migration

### PR DESCRIPTION
**Pull request by Suika-Kong**
Also bumping _format_version to 3.0.

Fix FT-3098

### CODE REVIEW

**Comment by Suika-Kong** [comment]
The migration should be done before validation.

**Comment by Suika-Kong** [comment]
Remember to downgrade the version number for 3.0/2.8 CP/DP

File: kong/templates/nginx_kong.lua [hunk 1]
```diff
@@ -439,7 +439,7 @@ server {
         }
     }
 
-> if legacy_hybrid_protocol == "off" then
+> if not legacy_hybrid_protocol then
```
Comment (fffonion) [inline kong/templates/nginx_kong.lua:442]: seems unrelated commits, needs a rebase?

Comment (Suika-Kong) [reply kong/templates/nginx_kong.lua]: done

File: kong/db/declarative/migrations/init.lua [hunk 2]
```diff
@@ -0,0 +1,12 @@
+local regex_route_path = require "kong.db.declarative.migrations.regex_route_path"
+
+return function (tbl)
+    if not tbl then
```
Comment (mayocream) [inline kong/db/declarative/migrations/init.lua:4]: Let's use 2 spaces ident to align with other files.

**Review by mayocream — COMMENTED** [review]
Let's add a test to make sure the 2.1 to 3.0 declarative yaml upgrading works.

**Review by mayocream — APPROVED** [review]

File: kong/db/migrations/migrate_regex_280_300.lua [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+  return (re_find(path, [[[a-zA-Z0-9\.\-_~/%]*$]], "ajo"))
+end
+
+local function migrate_regex(reg)
+  if is_not_regex(reg) then
+    return reg, false
+  end
+
+  local migrated = "~" .. normalize_regex(reg)
+  return migrated, migrated ~= reg
```
Comment (fffonion) [inline kong/db/migrations/migrate_regex_280_300.lua:83]: seems second return value is always true?
```suggestion
  return migrated, true
```

File: spec/01-unit/01-db/01-schema/11-declarative_config/04-on-the-fly-migration_spec.lua [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+  table.sort(tbl)
+  return tbl
+end
+
+
+describe("declarative config: on the fly migration", function()
+  it("routes handling", function()
+    local dc = assert(declarative.new_config(conf_loader()))
+    local config = [[
+      _format_version: "1.1"
```
Comment (fffonion) [inline spec/01-unit/01-db/01-schema/11-declarative_config/04-on-the-fly-migration_spec.lua:76]: make sure `2.1` is covered in tests, we can simply make it a loop and only change the version number

File: kong/db/migrations/migrate_regex_280_300.lua (hunk 3 above)
Comment (Suika-Kong) [reply kong/db/migrations/migrate_regex_280_300.lua:83]: Ah... that's a leftover of refactoring.

**Review by fffonion — APPROVED** [review]
